### PR TITLE
Add a module type for account validity

### PR DESCRIPTION
**Pull request by babolivier**
This PR adds an API for third-party plugin modules to implement account validity, so they can provide this feature instead of Synapse. The module implementing the current behaviour for this feature can be found at https://github.com/matrix-org/synapse-email-account-validity.

To allow for a smooth transition between the current feature and the new module, hooks have been added to the existing account validity endpoints to allow their behaviours to be overridden by a module.

~~Best reviewed commit by commit - there are quite a few of them but only the 4 first ones are a bit large.~~

~~Requires #9977~~

### CODE REVIEW

**Comment by babolivier** [comment]
I'd also appreciate if someone could sanity-check https://github.com/matrix-org/synapse-email-account-validity - it should be fairly OK since most of it is copy-pasted from the current account validity implem, but it'd be better if someone else has had a look at it.

File: UPGRADE.rst
Comment (richvdh) [inline UPGRADE.rst:97]: I think we'll need clearer instructions for admins here: how do they install the module, how should they update the config?

**Review by richvdh — CHANGES_REQUESTED** [review]
looks generally good, but a few bits and bpbs to fix up.

Thank you for trying to make it reviewable by splitting it up into separate commits, though there are a couple of things I saw in early commits which got fixed later, so excuse the irrelevant comments.

File: synapse/module_api/__init__.py
Comment (richvdh) [reply synapse/module_api/__init__.py:568]: (splitting up Mailer is probably best done in a preparatory PR)

File: UPGRADE.rst
Comment (babolivier) [reply UPGRADE.rst:97]: These instructions are available at https://github.com/matrix-org/synapse-email-account-validity, my idea was that admins would follow the link to the module and follow the steps described there, but maybe that's not clear enough?

Comment (richvdh) [reply UPGRADE.rst:97]: wellll... I think on the one hand, we need some migration instructions rather than instructions that start from scratch.

Also, `pip install` doesn't work for people who didn't install with pip. (docker, debian, etc, etc)

File: UPGRADE.rst
Comment (callahad) [inline UPGRADE.rst:99]: Should we define this more explicitly?

(E.g., giving people three months to migrate would mean dropping in Synapse 1.40 on 2021-08-10)

File: UPGRADE.rst
Comment (babolivier) [reply UPGRADE.rst:97]: > I think on the one hand, we need some migration instructions rather than instructions that start from scratch.

The module attempts a migration when starting fwiw so just installing it from scratch and replacing the `account_validity` config with the module's should be enough, but fair enough I'll add more instructions on this doc as well.

> Also, `pip install` doesn't work for people who didn't install with pip. (docker, debian, etc, etc)

True, however I don't know how we support non-pip installs since none of the other modules we publish seem to do.

File: synapse/api/auth.py
Comment (richvdh) [reply synapse/api/auth.py:228]: ohh, or better:

we change round the way modules work, so we just have a single `modules` list at the top level of the homeserver config. Each module gets loaded at startup, but after all the handlers. Then the module registers for the hooks it needs (in this case, `account_validity` and additional http resources).

Comment (babolivier) [reply synapse/api/auth.py:228]: > I'm guessing that we're doing this `hs.get_account_validity()` here rather than in the constructor to avoid a dependency loop?

Yes, iirc it's because `hs.get_auth()` is called in the module API's `__init__` function (or because a handler that's listed [here](https://github.com/matrix-org/synapse/blob/2dd3a354b1eb827c2ea54c075bfaa6aa942a1d45/synapse/server.py#L215-L223) does it).

> we change round the way modules work, so we just have a single `modules` list at the top level of the homeserver config. Each module gets loaded at startup, but after all the handlers. Then the module registers for the hooks it needs (in this case, `account_validity` and additional http resources).

I was thinking of that earlier, and I really like the idea. It would also have the added bonus of making documenting module development easier since you wouldn't need a different documentation for every different type of module.

Comment (richvdh) [reply synapse/api/auth.py:228]: >  I was thinking of that earlier, and I really like the idea.

Yeah. If you fancy making that change here, I think it will be very worthwhile (and hopefully not too much work).

Comment (babolivier) [reply synapse/api/auth.py:228]: I think I'm happy to do this change but given it changes some things about other modules as well it might be better to do it in a separate PR.

Comment (richvdh) [reply synapse/api/auth.py:228]: I think splitting up this PR is a good idea - it's got a bit unweildly.

I wouldn't leap into changing the way other modules work right now. We'll need to keep the existing interfaces around for quite a while, even if they are deprecated.

File: synapse/module_api/__init__.py
Comment (babolivier) [reply synapse/module_api/__init__.py:81]: Substitution fails if there are placeholders in the string that don't match a key in the dictionary:

```python
>>> "Your Friendly %(hs)s homeserver" % {"app": "foo"}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'hs'
>>> 
```

But yeah limiting to `KeyError` sounds like a sensible thing to do.

File: synapse/module_api/__init__.py
Comment (babolivier) [reply synapse/module_api/__init__.py]: (resolving as this got fixed in another commit, sorry for the faff)

File: synapse/module_api/__init__.py
Comment (babolivier) [reply synapse/module_api/__init__.py:591]: Thanks, I often get those confused

File: docs/sample_config.yaml
Comment (babolivier) [reply docs/sample_config.yaml:1395]: It's not 3 spaces though, it's 2 (4 total from the `#`)

Comment (richvdh) [reply docs/sample_config.yaml:1395]: there are definitely 3 spaces before the `#`.

Comment (babolivier) [reply docs/sample_config.yaml:1395]: Oh, I thought you meant the indent after the `#` :man_facepalming:

Comment (richvdh) [reply docs/sample_config.yaml:1395]: my bad, cryptic comment was cryptic

File: synapse/account_validity/__init__.py
Comment (babolivier) [reply synapse/account_validity/__init__.py:23]: I looked at other modules (namely spam checker and third party event rules) which have this code (instantiating + calling hooks on modules) in a dedicated class outside of handlers, and thought of the account validity handler as the legacy code that'll eventually go away.

File: docs/sample_config.yaml
Comment (babolivier) [reply docs/sample_config.yaml:1383]: Ah yes that makes more sense, thanks

File: synapse/module_api/__init__.py
Comment (babolivier) [reply synapse/module_api/__init__.py:570]: I've open https://github.com/matrix-org/synapse/pull/9977 for that.

File: synapse/api/auth.py
Comment (babolivier) [reply synapse/api/auth.py]: Resolving the thread as I believe the conclusion is that this will be done in a future PR (and I have the conn on making that happen)

File: UPGRADE.rst
Comment (babolivier) [reply UPGRADE.rst]: I'm a bit uncertain what this PR should be doing to fix this issue. We had a discussion about it in #synapse-dev and (I think?) came to a conclusion (I don't remember the specifics, but iirc it was about telling users how to build custom Docker images based on the Synapse one with the modules they want in it) but it feels a bit out of scope for this PR to me.

File: UPGRADE.rst
Comment (babolivier) [reply UPGRADE.rst:99]: Agreed that it needs defining more precisely. Given the time this PR is taking to get through, I'll determine that once the review is done since we'll at least know which version this change will be shipped in.

Comment (richvdh) [reply UPGRADE.rst:99]: historically we've always ended up looking stupid when we say "we will do X on date Y", and Y then sails past. It might be possible to come up with some wording that allows us wiggle room.

File: UPGRADE.rst
Comment (richvdh) [inline UPGRADE.rst:108]: ... which previous changes? I don't remember us doing this sort of thing before.

File: UPGRADE.rst
Comment (richvdh) [inline UPGRADE.rst:105]:
```suggestion
In line with previous changes around template management in Synapse, support has been
```

File: UPGRADE.rst
Comment (richvdh) [reply UPGRADE.rst]: I appreciate it feels like scope creep, but I think it's a prerequisite for moving functionality out of core Synapse into modules, which is why I questioned whether that was a good decision.

In short: if we're going down that path, I'd like us to have solutions in place for modules with docker *and debian* before this PR lands.

File: synapse/api/auth.py
Comment (richvdh) [reply synapse/api/auth.py]: I'm reopening this. Whilst we should leave the *existing* module hooks in place for a future PR, I don't think we should add any *new* module hooks that assume single-function modules, so I'm afraid I'd like to see changes before this PR lands.

File: UPGRADE.rst
Comment (babolivier) [reply UPGRADE.rst:108]: I seemed to remember we had done work to hardcode the name of templates in the past, but looking at other places in the code it looks like I've imagined that.

File: synapse/module_api/__init__.py
Comment (richvdh) [reply synapse/module_api/__init__.py]: yeah fair enough, I don't really have any great suggestions here. it's probably best to expose a `looping_call` method on ModuleApi which calls `run_as_background_process` on the function (I think you may as well call `run_as_background_process` directly rather than messing with `wrap_as_background_process`).

It might be nice to make the "description" an optional param rather than hardcoding it to the name of the function.

File: synapse/account_validity/__init__.py
Comment (richvdh) [reply synapse/account_validity/__init__.py:23]: hrm; I'm not convinced that's a great model. 

spam-checker seems to be an entire module to define three constants - I'm not really sure why we did that.

In my mind, the REST layer calls things in the handlers. It might be better to rename the existing `AccountValidityHandler` (to `LegacyAccountValidityImpl` or something) but keep it in the same file. Then you define a new `AccountValidityHandler` which is called by the REST layer, and manages the calls out to the module as well as the fallback to the `LegacyAccountValidityImpl`.

File: UPGRADE.rst
Comment (richvdh) [reply UPGRADE.rst:108]: as a general rule we have hardcoded names for new templates. Where we have existing config settings, we have removed them from the sample config, but they continue to work. I'd suggest doing the latter here.

likewise `template_dir`: I would remove it from the sample config, and have it default to the global `template_dir`, but keep it working if someone has deliberately set it.

File: UPGRADE.rst
Comment (babolivier) [reply UPGRADE.rst]: ftr we decided to ditch the deprecation issue for now, so this PR isn't blocked on finding a solution to this problem.

File: synapse/module_api/__init__.py
Comment (richvdh) [inline synapse/module_api/__init__.py:169]:
```suggestion
        """The application name configured in the homeserver's configuration.
```

File: synapse/module_api/__init__.py
Comment (richvdh) [inline synapse/module_api/__init__.py:167]:
```suggestion
    def email_app_name(self) -> str:
```

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py]: As I said on our call: I think it would be good to expand this a bit to explain the status of these "legacy" callbacks, why they are needed, and how/when we hope to get rid of them.

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py:121]: given that we're keeping the old code around for now anyway, do we really need all of these hooks? For instance, can't we have the core code handle requests to `/_synapse/admin/v1/account_validity/validity` and the module handle requests to `/_synapse/client/account_validity/validity`, thus making the `on_legacy_admin_request` hook redundant?

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py:36]:
```suggestion
# Temporary hooks to allow for a transition from `/_matrix/client` endpoints
# to `/_synapse/client/account_validity`. See `register_account_validity_callbacks`.
ON_LEGACY_SEND_MAIL_CALLBACK = Callable[[str], Awaitable]
ON_LEGACY_RENEW_CALLBACK = Callable[[str], Awaitable[Tuple[bool, bool, int]]]
ON_LEGACY_ADMIN_REQUEST = Callable[[Request], Awaitable]
```

**Review by richvdh — CHANGES_REQUESTED** [review]

File: docs/modules.md
Comment (babolivier) [reply docs/modules.md:219]: I feel like this will mislead people into thinking the function _has_ to end with `return None`

File: synapse/api/auth.py
Comment (babolivier) [reply synapse/api/auth.py:78]: Also worth noting that this isn't a new issue since the line just above gets the state handler (and the account validity handler was already called in this file before this change, the PR is merely moving where we get it from the `hs` object).

File: synapse/handlers/account_validity.py
Comment (babolivier) [reply synapse/handlers/account_validity.py:121]: I think it would be confusing from a user pov because e.g. if they receive an email but click on the link 1w later (e.g. because they were on holidays) and the switch from the core code to the module happens in between, the link wouldn't be doing anything (since it would only update the old database table). The only other way to mitigate this that I can think of would be to have both the core code and the module enabled at the same time, but that'd mean users would get 2 emails with 2 different links to click on at about the same time, which is probably quite confusing.

File: synapse/handlers/account_validity.py
Comment (babolivier) [reply synapse/handlers/account_validity.py]: I assume this has already been fixed by https://github.com/matrix-org/synapse/pull/9884/commits/f8754fed56e93de206c561877b6fdfc60fa9191c

File: synapse/module_api/errors.py
Comment (babolivier) [reply synapse/module_api/errors.py:18]: Because `get_user_by_req` raises it if the token is invalid or doesn't belong to an existing user. Which makes me realise the docstring for the `get_user_by_req` method on the module API is lying, will fix.

**Comment by richvdh** [comment]
fixes #10139

File: docs/modules.md
Comment (richvdh) [reply docs/modules.md:219]: if they do think that, it isn't the end of the world.

Right now it just looks like we've forgotten to document the return type, when compared with the other callback functions.

File: synapse/api/auth.py
Comment (richvdh) [reply synapse/api/auth.py:78]: I'm not seeing that, tbh. But it doesn't really matter.

File: synapse/handlers/account_validity.py
Comment (richvdh) [reply synapse/handlers/account_validity.py:122]: I'm not following you. We can handle `/_synapse/admin/v1/account_validity/validity` in the core while still having emails link to `/_synapse/client/account_validity/validity` ?

File: synapse/module_api/__init__.py
Comment (richvdh) [inline synapse/module_api/__init__.py:171]: it'd be nice to give this a return type annotation while you're changing it.

File: synapse/module_api/__init__.py
Comment (richvdh) [inline synapse/module_api/__init__.py]:
```suggestion
            *args: Positional arguments to pass to function.
            desc: The background task's description. Defaults to the function's name.

```

**Review by richvdh — CHANGES_REQUESTED** [review]
see the points above, also the tiny nit below.

File: synapse/api/auth.py
Comment (babolivier) [reply synapse/api/auth.py:78]: Oh yeah I was mixing up things, mb

File: synapse/handlers/account_validity.py
Comment (babolivier) [reply synapse/handlers/account_validity.py:122]: We can, but then we'd end up with two different tables (`account_validity`, and `email_account_validity` from https://github.com/matrix-org/synapse-email-account-validity) used for the same purpose but with divergent data.

Comment (richvdh) [reply synapse/handlers/account_validity.py:122]: ugh, I hadn't realised the new module was using a different table. Can it not use the same table?

Comment (babolivier) [reply synapse/handlers/account_validity.py:122]: I'm a bit wary of having an external module write to a table that belongs to Synapse's schema to be honest. Especially since it's tied to a feature we'll be getting rid of eventually; having an external module use it means we would need extra care deleting the feature and its storage component and imho it's very easy to just forget it and only remember when it breaks e.g. on a customer's infrastructure.

Plus the naming wouldn't be right, since it then wouldn't just be account validity, but rather one way of doing account validity (the email-based one), and other modules might implement other ways to verify the validity of an account (e.g. by checking with an auth backend).

Comment (richvdh) [reply synapse/handlers/account_validity.py:122]: hmmm yes, fair enough.

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py:113]: let's add some space for readability:
```suggestion
        # /_synapse/client/account_validity, because:
        #
```

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py:115]:
```suggestion
        #   * the way we register servlets means that modules can't register resources
```

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py:116]:
```suggestion
        #     under /_matrix/client
        #
```

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py:120]:
```suggestion
        # email-based account validity) to temporarily hook into the legacy endpoints so we
```

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py:330]: it'd be nice to update the docstring on this method to give some allusion to the fact this is to support the legacy renewal endpoint.

File: synapse/handlers/account_validity.py
Comment (richvdh) [inline synapse/handlers/account_validity.py:119]:
```suggestion
        #
        # Once the email-account-validity module is loaded, it will take control of account
        # validity by moving the rows from our `account_validity` table into its own table.
        # 
        # Therefore, we need to allow modules (in practice just the one implementing the
```

**Review by richvdh — APPROVED** [review]
lgtm otherwise

**Pull request merged**
